include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -0,0 +1,1386 @@
+void handleTwosided_interactionOnly(inout bool ts_flag, in instr_t instr)
+{
+#ifndef NO_TWOSIDED
+	if (!ts_flag && instr_getTwosided(instr))
+	{
+		ts_flag = true;
+		if (currInteraction.isotropic.NdotV<0.0)
+		{
+			flipCurrInteraction();
+		}
+	}
+#ifdef OP_BUMPMAP
+	ts_flag = instr_getOpcode(instr)==OP_BUMPMAP ? false:ts_flag;
+#endif
+#endif	
+}
+//call before executing an instruction/evaluating bsdf
+void handleTwosided(inout bool ts_flag, in instr_t instr, inout irr_glsl_LightSample s, inout irr_glsl_AnisotropicMicrofacetCache uf)
+{
+#ifndef NO_TWOSIDED
+	if (!ts_flag && instr_getTwosided(instr))
+	{
+		ts_flag = true;
+		if (currInteraction.isotropic.NdotV<0.0)
+		{
+			flipCurrInteraction();
+
+			s.NdotL = -s.NdotL;
+			s.TdotL = -s.TdotL;
+			s.BdotL = -s.BdotL;
+			uf.isotropic.NdotH = -uf.isotropic.NdotH;
+			uf.TdotH = -uf.TdotH;
+			uf.BdotH = -uf.BdotH;
+		}
+	}
+#ifdef OP_BUMPMAP
+	ts_flag = instr_getOpcode(instr)==OP_BUMPMAP ? false:ts_flag;
+#endif
+#endif
+}
+
+//TODO move or rename those
+//ggx
+float irr_glsl_ggx_height_correlated_cos_eval_DG(in float NdotH, in float NdotL, in float NdotV, in float a2)
+{
+	float NdotH2 = NdotH*NdotH;
+	float maxNdotL = max(NdotL, 0.0);
+	float NdotL2 = NdotL*NdotL;
+	float maxNdotV = max(NdotV,0.0);
+	float NdotV2 = NdotV*NdotV;
+
+	return irr_glsl_ggx_height_correlated_cos_eval_DG_wo_clamps(NdotH2, maxNdotL, NdotL2, maxNdotV, NdotV2, a2);
+}
+float irr_glsl_ggx_height_correlated_cos_eval_DG(in irr_glsl_LightSample s, in irr_glsl_IsotropicMicrofacetCache uf, in irr_glsl_IsotropicViewSurfaceInteraction i, in float a2)
+{
+	return irr_glsl_ggx_height_correlated_cos_eval_DG(
+		uf.NdotH,
+		s.NdotL,
+		i.NdotV,
+		a2
+	);
+}
+
+float irr_glsl_ggx_height_correlated_aniso_cos_eval_DG(in float NdotH, in float TdotH, in float BdotH, in float NdotL, in float TdotL, in float BdotL, in float NdotV, in float TdotV, in float BdotV, in float ax, in float ax2, in float ay, in float ay2)
+{
+	float NdotH2 = NdotH * NdotH;
+	float TdotH2 = TdotH * TdotH;
+	float BdotH2 = BdotH * BdotH;
+	float maxNdotL = max(NdotL, 0.0);
+	float NdotL2 = NdotL * NdotL;
+	float TdotL2 = TdotL * TdotL;
+	float BdotL2 = BdotL * BdotL;
+	float maxNdotV = max(NdotV, 0.0);
+	float NdotV2 = NdotV * NdotV;
+	float TdotV2 = TdotV * TdotV;
+	float BdotV2 = BdotV * BdotV;
+
+	return irr_glsl_ggx_height_correlated_aniso_cos_eval_DG_wo_clamps(NdotH2, TdotH2, BdotH2, maxNdotL, NdotL2, TdotL2, BdotL2, maxNdotV, NdotV2, TdotV2, BdotV2, ax, ax2, ay ,ay2);
+}
+float irr_glsl_ggx_height_correlated_aniso_cos_eval_DG(in irr_glsl_LightSample s, in irr_glsl_AnisotropicMicrofacetCache uf, in irr_glsl_AnisotropicViewSurfaceInteraction i, in float ax, in float ax2, in float ay, in float ay2)
+{
+	return irr_glsl_ggx_height_correlated_aniso_cos_eval_DG(
+		uf.isotropic.NdotH,
+		uf.TdotH,
+		uf.BdotH,
+		s.NdotL,
+		s.TdotL,
+		s.BdotL,
+		i.isotropic.NdotV,
+		i.TdotV,
+		i.BdotV,
+		ax, ax2, ay, ay2
+	);
+}
```

Comment:
I have these already?

Reply:
theres no builtins with struct parameters in our glsl lib

Reply:
there's no reason why you should be using `cos_Eval_DG` because you can handle beckmann + ggx and anisotropy in a way that reduces divergence... at the very least you should be using `wo_clamps` functions

Reply:
you wont need them https://github.com/buildaworldnet/IrrlichtBAW/pull/554/files#r526504743